src/main/scala/firrtl/passes/CheckCombLoops.scala

```diff
@@ -0,0 +1,146 @@
+
+  private case class LogicNode(name: String, inst: Option[String] = None, memport: Option[String] = None)
+
+  private def toLogicNode(e: Expression): LogicNode = e match {
+    case r: WRef =>
+      LogicNode(r.name)
+    case s: WSubField =>
+      s.exp match {
+        case modref: WRef =>
+          LogicNode(s.name,Some(modref.name))
+        case memport: WSubField =>
+          memport.exp match {
+            case memref: WRef =>
+              LogicNode(s.name,Some(memref.name),Some(memport.name))
+          }
+      }
+  }
+
+  private def getExprDeps(deps: mutable.Set[LogicNode])(e: Expression): Expression = e match {
+    case r: WRef =>
+      deps += toLogicNode(r)
+      r
+    case s: WSubField =>
+      deps += toLogicNode(s)
+      s
+    case _ =>
+      e map getExprDeps(deps)
+      e
+  }
+
+  private def getStmtDeps(
+    simplifiedModules: mutable.Map[String,DiGraph[LogicNode]],
+    deps: MutableDiGraph[LogicNode])(s: Statement): Statement = s match {
+    case c: Connect =>
+      val lhs = toLogicNode(c.loc)
+      if (deps.contains(lhs)) {
+        getExprDeps(deps.getEdges(lhs))(c.expr)
+      }
+      c
+    case w: DefWire =>
+      deps.addVertex(LogicNode(w.name))
+      w
+    case n: DefNode =>
+      val lhs = LogicNode(n.name)
+      deps.addVertex(lhs)
+      getExprDeps(deps.getEdges(lhs))(n.value)
+      n
+    case m: DefMemory if (m.readLatency == 0) =>
+      for (rp <- m.readers) {
+        val dataNode = deps.addVertex(LogicNode("data",Some(m.name),Some(rp)))
+        deps.addEdge(dataNode, deps.addVertex(LogicNode("addr",Some(m.name),Some(rp))))
+        deps.addEdge(dataNode, deps.addVertex(LogicNode("en",Some(m.name),Some(rp))))
+      }
+      m
+    case i: WDefInstance =>
+      val iGraph = simplifiedModules(i.module).transformNodes(n => n.copy(inst = Some(i.name)))
+      for (v <- iGraph.getVertices) {
+        deps.addVertex(v)
+        iGraph.getEdges(v).foreach { deps.addEdge(v,_) }
+      }
+      i
+    case _ =>
+      s map getStmtDeps(simplifiedModules,deps)
+      s
+  }
+
+  private def recoverCycle(
+    m: String,
+    moduleGraphs: mutable.Map[String,DiGraph[LogicNode]],
+    moduleDeps: Map[String, Map[String,String]], 
+    prefix: Seq[String],
+    cycle: Seq[LogicNode]): Seq[String] = {
+    def absNodeName(prefix: Seq[String], n: LogicNode) =
+      (prefix ++ n.inst ++ n.memport :+ n.name).mkString(".")
+    val cycNodes = (cycle zip cycle.tail) map { case (a, b) =>
+      if (a.inst.isDefined && !a.memport.isDefined && a.inst == b.inst) {
+        val child = moduleDeps(m)(a.inst.get)
+        val newprefix = prefix :+ a.inst.get
+        val subpath = moduleGraphs(child).path(b.copy(inst=None),a.copy(inst=None)).tail.reverse
+        recoverCycle(child,moduleGraphs,moduleDeps,newprefix,subpath)
+      } else {
+        Seq(absNodeName(prefix,a))
+      }
+    }
+    cycNodes.flatten :+ absNodeName(prefix, cycle.last)
+  }
+
+
+  def run(c: Circuit): Circuit = {
```

Comment:
Add an explanation of the algorithm (e.g. first topological sort, then iterate through modules, then check.